relayer/src/client.rs

```diff
@@ -5,116 +5,254 @@ use std::{
+        self.finalization_loop().await?;
+
+        Ok(())
+    }
 
-    exit_if_term(term.clone());
-
-    client.start().await?;
-    log::info!("client started");
-
-    let filter = ethers::types::Filter::new()
-        .select(
-            config
-                .block_number
-                .map(Into::into)
-                .unwrap_or(ethers::core::types::BlockNumber::Latest)..,
-        )
-        .address(config.smart_contract_address.parse::<Address>()?)
-        .event("Transfer(address,address,uint256)");
-
-    loop {
-        exit_if_term(term.clone());
-        let logs = client.get_logs(&filter).await?;
-        log::debug!("logs: {:#?}", logs);
-        'outer: for log in logs {
-            if let Some(block_hash) = log.block_hash {
-                if let Ok(Some(block)) = client.get_block_by_hash(&block_hash.encode(), false).await
-                {
-                    let mut receipts = vec![];
-                    for hash in transactions_to_hashes(block.transactions) {
-                        if let Ok(receipt) = client.get_transaction_receipt(&hash).await {
-                            receipts.push(receipt)
-                        } else {
-                            log::warn!(
-                                "Could not get a transaction receipt for tx {}",
-                                hash.encode_hex()
-                            );
-                            continue 'outer;
-                        }
-                    }
-
-                    if !receipts.is_empty() {
-                        let json = serde_json::to_string(&receipts)?;
-                        db.insert_receipts(block_hash, &json)?;
-                    } else {
-                        log::debug!(
-                            "Block {} does not have any receipts",
-                            block_hash.encode_hex()
-                        );
-                    }
-                } else {
-                    log::info!(
-                        "Could not get a block by block_hash {}",
-                        block_hash.encode_hex()
-                    );
-                    continue 'outer;
-                }
+    async fn finalization_loop(&mut self) -> Result<()> {
+        let mut latest_finalized_block =
+            self.db.select_latest_block_height(BlockType::Finalized)?;
+        let mut duration = tokio::time::interval(Duration::from_secs(5));
+        loop {
+            exit_if_term(self.term.clone());
+            duration.tick().await;
+            let finalized_block = self
+                .client
+                .get_block_by_number(BlockTag::Finalized, false)
+                .await;
+            let finalized_block = if let Ok(Some(finalized_block)) = finalized_block {
+                finalized_block
+            } else {
+                log::warn!(target: "relayer::client::finalization_loop","Failed to get finalized block, retrying in {} seconds", duration.period().as_secs());
```

Comment:
You can define `let target = "relayer::client::finalization_loop";` and use it in all `log::*!(...)`

Reply:
done